lib/CustomFields/components/CustomFieldsForm/CustomFieldsForm.js

```diff
@@ -0,0 +1,361 @@
+  entityType: PropTypes.string.isRequired,
+  fetchUsageStatistics: PropTypes.func.isRequired,
+  initialValues: PropTypes.arrayOf(PropTypes.shape({
+    id: PropTypes.string.isRequired,
+    values: PropTypes.oneOfType([
+      textboxShape,
+    ]).isRequired,
+  })).isRequired,
+  onSubmit: PropTypes.func.isRequired,
+  redirectToView: PropTypes.func.isRequired,
+  submitSucceed: PropTypes.bool.isRequired,
+  submitting: PropTypes.bool.isRequired,
+};
+
+const CustomFieldsForm = ({
+  submitSucceed,
+  submitting,
+  fetchUsageStatistics,
+  redirectToView,
+  initialValues,
+  entityType,
+  onSubmit,
+}) => {
+  const getInitialAccordionsState = () => {
+    return initialValues.reduce((state, cf) => ({
+      ...state,
+      [cf.id]: false,
+    }), {});
+  };
+
+  const [accordions, setAccordions] = useState(getInitialAccordionsState);
+
+  const getInitialFormState = () => {
+    return {
+      customFields: initialValues,
+      pristine: true,
+      fieldsToDelete: [],
+    };
+  };
+
+  const stateReducer = (state, { type, payload }) => {
+    switch (type) {
+      case 'ADD_NEW_FIELD': {
+        const updatedCustomFields = [...state.customFields];
+        updatedCustomFields.push({
+          values: {
+            ...defaultFieldConfigs[payload.type],
+            entityType,
+          },
+          id: uniqueId('_unsaved'),
+        });
+
+        return {
+          ...state,
+          customFields: updatedCustomFields,
+          pristine: false,
+        };
+      }
+      case 'UPDATE_FIELD_VALUE': {
+        const fieldIndex = state.customFields.findIndex(cf => payload.fieldId === cf.id);
+
+        const updatedCFData = {
+          ...state.customFields[fieldIndex],
+          values: {
+            ...state.customFields[fieldIndex].values,
+            ...payload.data,
+          }
+        };
+
+        const updatedCustomFields = [...state.customFields];
+        updatedCustomFields[fieldIndex] = updatedCFData;
+
+        return {
+          ...state,
+          customFields: updatedCustomFields,
+          pristine: isEqual(updatedCustomFields, initialValues),
+        };
+      }
+      case 'DELETE_CUSTOM_FIELD': {
+        const newCustomFields = [...state.customFields];
+        const currentFieldPosition = newCustomFields.findIndex(cf => cf.id === payload.fieldId);
+        const deletedField = newCustomFields.splice(currentFieldPosition, 1)[0];
+        const updatedFieldsToDelete = [...state.fieldsToDelete];
+
+        if (!deletedField.id.startsWith('_unsaved')) {
+          updatedFieldsToDelete.push({ index: currentFieldPosition, data: deletedField });
+        }
+        return {
+          customFields: newCustomFields,
+          pristine: isEqual(newCustomFields, initialValues),
+          fieldsToDelete: updatedFieldsToDelete,
+        };
+      }
+      case 'CANCEL_DELETE': {
+        const newCustomFields = [...state.customFields];
+        state.fieldsToDelete.forEach(cf => {
```

Comment:
I think it would be better to use there filter. No need to change if you think otherwise.
```suggestion
   const newCustomFields =  state.customFields.filter(customField => !state.fieldsToDelete.includes(cf.index))
```